network/src/dms/mod.rs

```diff
@@ -0,0 +1,339 @@
+                .await
+                .read_file(&format!("metadata-{}", &file_name[8..]))
+                .await
+                .map(|metadata| (metadata, message))
+        });
+        let messages = future::join_all(tasks)
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut result = Vec::new();
+        for (metadata, message) in &messages {
+            let metadata = serde_spb::from_str::<MessageMetadata>(metadata)
+                .map_err(|e| IntegrityError::new(format!("can't decode stored data: {e}")))?;
+            let message = serde_spb::from_str::<M>(message)
+                .map_err(|e| IntegrityError::new(format!("can't decode stored data: {e}")))?;
+            result.push((message, metadata));
+        }
+        Ok(result)
+    }
+
+    fn test_membership(&self, member: &PublicKey) -> bool {
+        self.config.members.contains(member)
+    }
+
+    async fn receive_packet(&mut self, packet: Packet) -> Result<(), Error> {
+        let message = serde_spb::from_slice::<M>(&packet.message)?;
+        message.verify_commitment(&packet.commitment, &self.config.dms_key)?;
+        if !self.test_membership(&packet.commitment.committer) {
+            return Err(eyre!("commitment committer is not a member"));
+        }
+        self.store_message(&message, packet.commitment).await?;
+        Ok(())
+    }
+
+    async fn store_message(
+        &mut self,
+        message: &M,
+        commitment: MessageCommitmentProof,
+    ) -> Result<(), Error> {
+        let message_hash = message.to_hash256();
+        if let Some((_, mut metadata)) = self.read_raw_message(message_hash).await? {
+            if metadata.committers.contains(&commitment) {
+                return Ok(());
+            } else {
+                metadata.committers.push(commitment);
+                self.storage
+                    .write()
+                    .await
+                    .add_or_overwrite_file(
+                        &format!("metadata-{message_hash}.json"),
+                        serde_spb::to_string(&metadata).unwrap(),
+                    )
+                    .await?;
+            };
+        } else {
+            let mut storage = self.storage.write().await;
+            storage
+                .add_or_overwrite_file(
+                    &format!("metadata-{message_hash}.json"),
+                    serde_spb::to_string(&MessageMetadata {
+                        message_hash,
+                        committers: vec![commitment],
+                    })
+                    .unwrap(),
+                )
+                .await?;
+            storage
+                .add_or_overwrite_file(
+                    &format!("message-{message_hash}.json"),
+                    serde_spb::to_string(&message).unwrap(),
+                )
+                .await?;
+        };
+        Ok(())
+    }
+
+    async fn retrieve_packtes(&self) -> Result<Vec<Packet>, Error> {
```

Comment:
Need to be refactored!
(packtes -> packets)

Reply:
fixed.